### PR TITLE
GUI: Add Python interface to create custom actions

### PR DESCRIPTION
**Pull request by chennes**
Add two static Python commands to FreeCADGui.Command enabling Python code to add and remove custom macro commands.

@wwmayer I am not sure these are in the right place, or if there's a better place or way to achieve this -- I'd appreciate it if you could take a look. This is a bit of copy-and-paste work, I'm not that confident in my Python binding coding.

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5501">pipeline status</a> for feature branch [PR_5501](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5501). Pipeline [464176162](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/464176162) was triggered at [15f8cd7](https://github.com/FreeCAD/FreeCAD/pull/5501/commits/15f8cd7724b1dff38a529aaf90fdd2dbc799e355). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5501">pipeline status</a> for feature branch [PR_5501](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5501). Pipeline [464178241](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/464178241) was triggered at [c298e65](https://github.com/FreeCAD/FreeCAD/pull/5501/commits/c298e650b603c786290482e88210b8c5b5afaa68). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Review by yorikvanhavre — APPROVED** [review]
Highly welcome! On our way to auto-installable macros? :)
The place looks fine to me, as your python binding code. Maybe later a listActions() would also be convenient?
Only small question, the term "Action" is new in FreeCAD slang (although it's well-known among Qt developers), maybe something like "Macro Shortcut" would be more in sync with the FreeCAD UI? Not really a strong opinion...

**Review by wwmayer — CHANGES_REQUESTED** [review]
See comments below.

File: src/Gui/CommandPyImp.cpp [hunk 1]
```diff
@@ -321,6 +321,92 @@ PyObject* CommandPy::getAction(PyObject *args)
...
+    if (pixmapTxt)
+        macro->setPixmap(pixmapTxt);
+
+    if (shortcutTxt)
+        macro->setAccel(shortcutTxt);
+
+    return PyUnicode_FromString(name);
+}
+
+PyObject* CommandPy::removeCustomAction(PyObject* args)
```
Comment (wwmayer) [inline src/Gui/CommandPyImp.cpp:388]: Same here

File: src/Gui/CommandPyImp.cpp [hunk 2]
```diff
@@ -321,6 +321,92 @@ PyObject* CommandPy::getAction(PyObject *args)
     }
 }
 
+// Behavior mimicked from DlgActionsImp.cpp
+QString newActionName()
```
[the remark this answers is not in the document]
Comment (chennes) [reply src/Gui/CommandPyImp.cpp]: Done - also then refactored DlgActionsImp to use the new function.

File: src/Gui/CommandPyImp.cpp [hunk 3]
```diff
@@ -321,6 +321,92 @@ PyObject* CommandPy::getAction(PyObject *args)
...
+                used = true;
+                break;
+            }
+        }
+    }
+
+    return name;
+}
+
+PyObject* CommandPy::createCustomAction(PyObject* args)
```
[the remark this answers is not in the document]
Comment (chennes) [reply src/Gui/CommandPyImp.cpp]: Done, both commands are renamed "*Command" instead of "*Action".

**Comment by chennes** [comment]
Thank you both for the review: I've made the suggested changes. I'll leave this here for a couple days in case you want to have another look.

**Comment by wwmayer** [comment]
Nice, thank you! The code looks good to me and you can merge it if you want.

**Comment by luzpaz** [comment]
CC @galou thought you'd dig this (and maybe something to keep in mind over at https://github.com/FreeCAD/FreeCAD-macros/ ?)

**Comment by galou** [comment]
@luzpaz you're right, I'll add the required pieces of information to the metadata: `Menu text`, `Tool tip`, `Status text`, and `What's this` as I read in the `Customize` dialog. @chennes which ones of those do you actually use?

I understand them like this:

- "Menu text" is shown if there is no icon
- "Tool tip" is shown at the mouse position when you hover the mouse over the toolbar button
- "Status text" is shown in the status bar when you hover the mouse over the toolbar button
- "What's this" should be shown when you click on the toolbar button after having clicked on the `What's this` button but it doesn't work on FreeCAD daily (`FreeCAD help files not found (/usr/share/doc/freecad-daily-doc/freecad.qhc). You might need to install the FreeCAD documentation package.`)

Am I correct?

**Pull request merged**